DHIS2: Creating and updating cases

kaapstorm opened the pull request:
Fixed program enrollment for Nutrition Assessment -- DHIS2 complaining about values, but not identifying which values, or what was wrong with them. 

Moved the creating and updating of cases into their own functions

Changed the way org units are iterated

kaapstorm committed:
DHIS2: Removed unused import

kaapstorm committed:
DHIS2: Treat DHIS2 settings as a dict

kaapstorm committed:
Merge branch 'master' into dhis2

kaapstorm committed:
DHIS2: Better query to iterate enabled domains

kaapstorm committed:
DHIS2: Manage lookup tables better

kaapstorm committed:
DHIS: Fixtures' "name" and "parent_id" are attributes, not items.

kaapstorm committed:
DHIS2: Added logging

kaapstorm committed:
DHIS2: Moved case type and TE name to constants

kaapstorm committed:
DHIS2: Fixed enrolling

kaapstorm committed:
DHIS2: Add and update cases in separate functions

kaapstorm committed:
DHIS2: Add org units as you iterate them

kaapstorm committed:
Merge branch 'master' into dhis2

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS: Iterate org units faster

Only get org unit details if you need them

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Use case domain instead of case owner project

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Use program fields for cases, not event elements. And DRY.

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Cache top org unit

kaapstorm committed:
DHIS2: Move _data_to_attributes to its own method

kaapstorm committed:
DHIS2: Avoid DHIS2 2.16 database bug

kaapstorm committed:
DHIS2: Don't sync org units; prod has over 10 000

kaapstorm committed:
DHIS2: Fix UserES query

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
Merge branch 'master' into dhis2

kaapstorm committed:
DHIS2: Cater for Risk Assessment-style data (i.e. dropdowns)

kaapstorm committed:
Always associate events with cases

kaapstorm committed:
DHIS2: Sync org untis after all

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Use form forwarding to push data to DHIS2

kaapstorm committed:
DHIS2: Fixed typo

kaapstorm committed:
DHIS2: Added logging to payload generator, fixed AttributeError

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
Merge branch 'master' into dhis2

kaapstorm committed:
DHIS2: Fixed bug getting case of form

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Massage API return values to look consistent

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Only update attributes we've got

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: "Event Date" event data element not found

kaapstorm committed:
DHIS2: Misspelled constants

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Skip risk assessment checkbox values that aren't checked

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Always include program stage

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Added view for testing form forwarding on Staging

dimagimon commented on custom/dhis2/views.py:
```diff
@@ -0,0 +1,28 @@
+from datetime import datetime
+from corehq.apps.receiverwrapper.models import RepeatRecord
+from corehq.apps.receiverwrapper.tasks import process_repeater_list
+from custom.dhis2.models import Dhis2Settings
```
F401 'Dhis2Settings' imported but unused

kaapstorm committed:
Merge branch 'master' into dhis2

kaapstorm committed:
DHIS2: Updated comments

kaapstorm commented:
I think DHIS2 is ready for a final merge.

I've added a test page to check repeat records at https://[hostname]/a/[domain]/dhis2/check/ so that forms can be forwarded to DHIS2 from Staging (or immediately from localhost).

I'd like to update the mocks and extend the automated tests, but that can go under future work, when we make this more generic than just wv-lanka.

kaapstorm committed:
DHIS2: Unused import

dimagimon commented:
:+1: No lint errors found.

snopoke commented on custom/dhis2/models.py:
```diff
@@ -179,63 +242,58 @@ def add_te_inst(self, entity_data, te_name, ou_id):
-            'attributes': attributes
+            'attributes': self._data_to_attributes(instance_data)
         }
         __, response = self._request.post('trackedEntityInstances', request_data)
```
Doesn't seem like you ever check the status code returned by `JsonApiRequest`. Maybe consider just not returning it?

kaapstorm replied:
That's a really good point.

I think I like status codes more than I should. I'm not sure why. (And they make less sense with DHIS2 because sometimes it returns 500s (something the server did) when it should return 400s (something the client did).) I'll try to wean myself off them. :-)

snopoke commented on custom/dhis2/tasks.py:
```diff
@@ -161,31 +146,68 @@ def get_user_by_org_unit(domain, org_unit_id, top_org_unit_name):
-        return CommCareUser.wrap(result.hits[i])
+    # user_data is a dynamic mapping. Despite documentation ...
+    # http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/mapping-dynamic-mapping.html
+    # ... it seems not to be filtering on it. We have to do this ourselves.
```
I think we're on an earlier release of ES

snopoke commented on custom/dhis2/views.py:
```diff
@@ -0,0 +1,28 @@
+</html>
+"""
+
+
```
This should get a `@domain_admin_required` decorator.

kaapstorm replied:
Is Sheel a superuser? Will he be able to use this?

snopoke replied:
Yes, pretty much all Dimagi users are super users.

snopoke commented on custom/dhis2/tasks.py:
```diff
@@ -214,28 +236,22 @@ def gen_children_only_ours(domain):
+# There is a large number of org units, but the lookup table is not deployed to handsets.
 @periodic_task(run_every=crontab(minute=3, hour=3))  # Run daily at 03h03
-def sync_org_units():
+def fetch_org_units():
```
I thought we weren't going to sync the org units.

kaapstorm replied:
Apparently they don't get sent to handsets. Sheel wants the list available in HQ to look up org unit IDs.

snopoke replied:
Oh right, I see there is a setting to say if the fixture is 'visible to all
users'

kaapstorm committed:
DHIS2: domain_admin_required for check_repeaters view

kaapstorm committed:
DHIS2: Drop unused return value "status_code"

kaapstorm committed:
DHIS2: Corrected comment

kaapstorm committed:
DHIS2: require_superuser for check_repeaters view

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Doesn't like ISO-formatted timestamps

dimagimon commented:
:+1: No lint errors found.

snopoke commented:
Looks good, waiting for build.

kaapstorm committed:
DHIS2: Fixed tests

dimagimon commented:
:+1: No lint errors found.

kaapstorm committed:
DHIS2: Iterate 500 org units per page

kaapstorm committed:
DHIS2: Massage Gender values before sending to DHIS2

kaapstorm committed:
DHIS2: Raise exception if unable to add an entity instance

kaapstorm committed:
DHIS2: Handle questions with multiple checkboxes correctly

kaapstorm committed:
Merge branch 'master' into dhis2

kaapstorm committed:
DHIS2: Include case_name when importing from DHIS2

Also just map child_id straight CHDR Number

kaapstorm committed:
Use IgnoreDocument exception to clean up form repeater queue

If get_payload returns None, HTTP post will fail, and retry 3 times.
Instead of making unnecessary requests, and clogging queue, just ignore
the document.

kaapstorm committed:
Revert "Use IgnoreDocument exception to clean up form repeater queue"

This reverts commit 81015455bac6a8e8947b2322ffde2d932dd48d1e.

Moved change to ignore_doc_exception branch. I think it complicates the DHIS2
branch, which is probably ready for merging.
